Clarify that partial verification is a minimum (#165)
uptane/uptane-standard: uptane-standard.md

#### OEMs initializing Uptane at the factory using SOTA

An OEM plans to install Uptane on new vehicles. This entails the following components: code to perform full and partial verification, the latest copy of the relevant metadata, the public keys, and an accurate attestation of the latest time. The OEM then either requires its tier-1 suppliers to provide these materials to the suppliers' assembly lines or can choose to add the materials later at the OEM's assembly lines. The OEM's in-vehicle implementation is Uptane-compliant if:

1. all Primaries perform full verification;
1. all Secondaries that are updated via OTA at least perform partial verification; and

> 1. all other ECUs that do not perform verification cannot be updated via OTA.

Reviewer: ```suggestion
1. all other ECUs that do not perform any type of verification cannot be updated via OTA.
```

An OEM plans to install Uptane on new vehicles. This entails the following components: code to perform full and partial verification, the latest copy of the relevant metadata, the public keys, and an accurate attestation of the latest time. The OEM then either requires its tier-1 suppliers to provide these materials to the suppliers' assembly lines or can choose to add the materials later at the OEM's assembly lines. The OEM's implementation is Uptane-compliant if:

1. all primaries perform full verification;

> 1. all secondaries that are updated via OTA perform partial verification at a minimum; and

Reviewer 2: A little bit vague, because there's only minimum (partial) and maximum (full).

Reviewer: I think this is a question of emphasis @trishankatdatadog. We want to establish partial verification as a required baseline. If you think the word "minimum" is confusing in this context, maybe it should be phrased "all secondaries that  are updated via OTA MUST at least perform partial verification" ..

Reviewer: ```suggestion
1. all secondaries that are updated via OTA MUST at least perform partial verification; and
```

Reviewer: Here again...I think this answers the concern

> 1. all secondaries that are updated via OTA MUST at least perform partial verification; and

Reviewer 3: Normally "Primary" and "Secondary" should be capitalized according to our style guide. I must've missed these because they are plural. Anyway, not critical and not the fault of your PR.

Author: We can't add a MUST until 2.0, but I applied the suggestion without the MUST in it.

An Uptane-compliant ECU SHALL be able to download and verify Image metadata and image binaries before installing a new image and MUST have a secure way of verifying the current time, or a sufficiently recent attestation of the time.

Each ECU receiving over-the-air updates in a vehicle is either a Primary or a Secondary ECU. A Primary ECU collects and delivers vehicle manifests to the Director ({{vehicle_version_manifest}}) that contain information about which images have been installed on ECUs in the vehicle. It also verifies the time and downloads and verifies the latest metadata and images for itself and for its Secondaries. A Secondary ECU verifies the time, and downloads and verifies the latest metadata and images for itself from its associated Primary ECU. It also sends signed information about its installed images to its associated Primary.

> All ECUs MUST verify image metadata as specified in {{metadata_verification}} before installing an image or making it available to other ECUs. A Primary ECU MUST perform full verification ({{full_verification}}). A Secondary ECU SHOULD perform full verification if possible. If a Secondary cannot perform full verification, it SHALL, at the very least, perform partial verification. In addition, it MAY also perform some steps from the full verification process. See the Uptane Deployment Considerations ({{DEPLOY}}) for a discussion of how to choose between partial and full verification.

Reviewer 2: Hmm... what "additional steps"? I don't remember. Is it in Deployment Considerations?

Seems to be related to [this](https://github.com/uptane/uptane-standard/pull/165#discussion_r437591845), but just needs tying it all up together.

Author: I think explaining the "why" is what the deployment considerations doc is for; let's leave this for there.

There might be different reasons to choose to implement more checks; one obvious one is to check root metadata from director so you don't need a recall/out-of-band update if you need to rotate director's targets key.

Each ECU in a vehicle receiving over-the-air updates is either a Primary or a Secondary ECU. A Primary ECU collects and delivers vehicle manifests to the Director ({{vehicle_version_manifest}}) that contain information about which images have been installed on ECUs in the vehicle. It also verifies the time and downloads and verifies the latest metadata and images for itself and for its secondaries. A Secondary ECU verifies the time, and downloads and verifies the latest metadata and images for itself from its associated Primary ECU. It also sends signed information about its installed images to its associated Primary.

> All ECUs MUST verify image metadata as specified in {{metadata_verification}} before installing an image or making it available to other ECUs. A Primary ECU MUST perform full verification ({{full_verification}}). A Secondary ECU SHOULD perform full verification if possible. If a Secondary cannot perform full verification, it SHALL perform partial verification at a minimum, but MAY perform any additional subset of full verification. See the Uptane Deployment Considerations ({{DEPLOY}}) for a discussion of how to choose between partial and full verification.

Reviewer 2: How do you "perform any additional subset of full verification"?

Reviewer: ```suggestion
All ECUs MUST verify image metadata as specified in {{metadata_verification}} before installing an image or making it available to other ECUs. A Primary ECU MUST perform full verification ({{full_verification}}). A Secondary ECU SHOULD perform full verification if possible. If a Secondary cannot perform full verification, it SHALL, at the very least, perform partial verification. In addition, it MAY also perform some steps from the full verification process. See the Uptane Deployment Considerations ({{DEPLOY}}) for a discussion of how to choose between partial and full verification.
```

Reviewer: I don't know if this quite fixes the wording problem, but I think its a step closer.

#### Send metadata to Secondaries {#send_metadata_primary}

> The Primary SHALL send its latest downloaded metadata to all of its associated Secondaries. The metadata it sends to each Secondary MUST include all of the metadata required for verification on that Secondary. For full verification Secondaries, this includes the metadata for all four roles from both repositories, plus any delegated Targets metadata files the Secondary will recurse through to find the proper delegation. For partial verification Secondaries, this MAY include fewer metadata files; it includes only the Targets metadata file from the Director repository at a minimum.

Reviewer 2: Do partial verification secondaries have a mechanism to rotate Targets key from Director? I forget. If yes, then we will need to clarify.

### Metadata verification procedures {#metadata_verification}

> A Primary ECU MUST perform full verification of metadata. A Secondary ECU SHOULD perform full verification of metadata. If a Secondary cannot perform full verification, it SHALL perform at least partial verification instead.

Reviewer 2: Same reservation about "at least": there's either partial or full, I don't understand what a subset would look like.

Reviewer: See explanation above. We are trying to establish that there is a "minimum" that must be performed. If that point is not clear, perhaps it needs to be rephrased.

Reviewer: It's the word order in this case. See suggestion:
```suggestion
A Primary ECU MUST perform full verification of metadata. A Secondary ECU SHOULD perform full verification of metadata. If a Secondary cannot perform full verification, it SHALL, at the very least, perform partial verification.
```

1. Load and verify the current time or the most recent securely attested time.
2. Download and check the Targets metadata file from the Director repository, following the procedure in {{check_targets}}.

Note that this verification procedure is intended to be a minimum. A partial verification ECU MAY implement more metadata checks.

> For example, in many cases it is desirable to also fetch and verify Root metadata from the Director (following the procedure in {{check_root}}) before checking Targets metadata. Without this check, an ECU will not be able to verify any metadata from the Director if the Director's Targets key is rotated, possible necessitating a recall.

Reviewer 2: Ok, I see my question has been answered here. Needs to be linked to my question above, IMO.

Reviewer: ```suggestion
Note that this verification procedure outline the minimal number of actions that must be taken. A partial verification ECU MAY implement more metadata checks.
```

Reviewer: I can't insert the suggestion in the right place for some reason. But, I would re-word line 832 as:

Note that this verification procedure outline the minimum number of actions that MUST be performed.

> For example, in many cases it is desirable to also fetch and verify Root metadata from the Director (following the procedure in {{check_root}}) before checking Targets metadata. Without this check, an ECU will not be able to verify any metadata from the Director if the Director's Targets key is rotated, possible necessitating a recall.

Reviewer 3: ```suggestion
For example, in many cases it is desirable to also fetch and verify Root metadata from the Director (following the procedure in {{check_root}}) before checking Targets metadata. Without this check, an ECU will not be able to verify any metadata from the Director if the Director's Targets key is rotated, possibly necessitating a recall.
```
